Fix unreadable source dirs

rgrinberg opened the pull request:
This PR makes it so that unreadable directories are treated as empty. Any time an unreadable dir is encountered, we report this as a warning.

One thing that would have been nicer would be to completely exclude them from the contents, but that breaks laziness a bit. Because we have to try reading the dir before adding it to the map.

rgrinberg committed:
Make File_tree.t tolerant to permission errors

Any node that fails to be read will now just emit a warning and be otherwise
ignored in the build.

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

ejgallego reviewed (commented):
Commenting to remove my review request [which seems spurious]

diml commented on src/stdune/file.mli:
```diff
@@ -0,0 +1,12 @@
+type t =
```
If we move this module to a more  central place, its name needs to identify more universally what this module is about. `File` doesn't seem like the right name to me.

rgrinberg replied:
Yeah the name isn't great. Happy to take suggestions

diml replied:
I don't have a better suggestions apart from something long and descriptive such as `Inode_and_device_number`. Given the size of this module, it makes me think that we should keep it as a sub-module for now.

rgrinberg replied:
Indeed. I moved it back.

rgrinberg commented:
I also made our symlink check warn instead of error. That seems to be a bit more friendly.

rgrinberg commented:
I've reverted to the old behavior regarding symlinks. It's not working in CI for some reason.

rgrinberg commented:
@diml this is ready by the way

diml commented on src/file_tree.ml:
```diff
@@ -127,77 +169,90 @@ type t = Dir.t
+            end)
+          |> Result.ok
+      ) in
+      let load_status = lazy (
```
This seems unnecessarily complicated to me. Do we really need all these level of lazy? In particular, it seems to me that we never force only one of `files` or `unfiltered_sub_dirs`, which makes me thing that it is not necessary to return them as separate lazy values.

rgrinberg replied:
To read the project file you only need files actually. Here I disagree that laziness complicates things as its just a flat list of bindings.  But since it doesn't buy us all that much I can get rid of it.

diml replied:
In practice, I don't think we can be in a situation where we would read the dune-project file and not recurse through the sub-directories. So getting rid of it seems best to me, at least it simplifies the code

rgrinberg replied:
Yeah, but there are certainly cases where we don't need to sort all the files. For big directories, that might not be so small.

rgrinberg commented on src/file_tree.ml:
```diff
@@ -127,77 +169,90 @@ type t = Dir.t
+    let load_status, files, unfiltered_sub_dirs =
+      let combined = lazy (
+        match Path.readdir_unsorted path with
+        | exception exn ->
```
Done

rgrinberg commented:
One concrete situation where the extra laziness helped was to avoid sorting directories when they were all ignored. That seems like an edge case though.

Now that we're only catching Unix errors, I've changed the result type to `(_, Unix.error) Result.t`. Using `Or_exn.t` is confusing since the function still raises in some situations (if Path.stat raises for example).

rgrinberg commented:
@emillon or @aalekseyev, mind taking over the review?

rgrinberg commented:
I'll add a test here by the way.

rgrinberg commented:
By the way, this is the last PR that I'd like to get in for 1.9

diml commented:
Ok, I had another look. This looks good, but why did you make the loading of dune-project files lazy? In practice, it means that errors in dune-project files might not be reported as eagerly as before which is not great.

Apart from this, it looks good to me.

rgrinberg commented:
@diml I removed the laziness and I've also added another feature:

If the directory is not readable, it will not be added as a node to `File_tree.t`. This costs us some laziness, but I think it's quite error prone to include such nodes in `File_tree` and have them be accessible via `fold` for example.

diml commented:
Ok, that seems good

rgrinberg commented:
@diml your last change makes it so that `Dune_project.anonymous` is always forced. Not a big deal, but it's the reason why I kept that extra bit of laziness.

diml commented:
Ok. Yh it doesn't matter in practice, the only reason `anonymous` is lazy is because it needs to be evaluated after the `"dune"` language is registered. Computing it is effectively fast and not worth avoiding.
